src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotUseReferenceEqualsWithValueTypes.cs

```diff
@@ -0,0 +1,95 @@
+
+        internal static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Reliability,
+            RuleLevel.BuildWarning,
+            description: s_localizableDescription,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.EnableConcurrentExecution();
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            analysisContext.RegisterCompilationStartAction(compilationStartContext =>
+            {
+                var objectType = compilationStartContext.Compilation.GetSpecialType(SpecialType.System_Object);
+
+                // Without these symbols the rule cannot run
+                if (objectType == null)
+                {
+                    return;
+                }
+
+                var referenceEqualsMethodGroup = objectType.GetMembers("ReferenceEquals").OfType<IMethodSymbol>();
+                var referenceEqualsMethod = referenceEqualsMethodGroup.GetFirstOrDefaultMemberWithParameterInfos(
+                    ParameterInfo.GetParameterInfo(objectType),
+                    ParameterInfo.GetParameterInfo(objectType));
+
+                if (referenceEqualsMethod == null)
+                {
+                    return;
+                }
+
+                compilationStartContext.RegisterOperationAction(operationContext =>
+                {
+                    var invocationExpression = (IInvocationOperation)operationContext.Operation;
+                    var targetMethod = invocationExpression.TargetMethod;
+
+                    if (!referenceEqualsMethod.Equals(targetMethod))
+                    {
+                        return;
+                    }
+
+                    foreach (var argument in invocationExpression.Arguments)
+                    {
+                        var val = argument.Value;
+
+                        // Only check through one level of conversion,
+                        // which will be either the boxing conversion to object,
+                        // or a reference type implicit conversion to object.
+                        if (val is IConversionOperation conversion)
+                        {
+                            val = conversion.Operand;
+                        }
+
+                        if (val.Type?.IsValueType == true)
+                        {
+                            operationContext.ReportDiagnostic(
+                                val.CreateDiagnostic(
+                                    Rule,
+                                    val.Type.ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat)));
```

Comment:
We cannot use this error message format because the analyzer is supported for both VB and C#. Generally we just use the simple `type.Name` in messages.

Reply:
I was concerned about that, but a lot of analyzers seem to use that format, none use the VB format, and the VB tests say that it says "Integer" instead of "int", so whatever the printer is it seems to understand which language applies.